### PR TITLE
Issue #7739: Update AbstractChecks to log DetailAST - JavadocMethod

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #7739: Update AbstractChecks to log DetailAST - JavadocMethod

this PR is based on comments in Issue #7739 and PR #8214  

diff report: https://huganghui.github.io/7739-JavadocMethod//diff/

### CODE REVIEW

**Comment by HuGanghui** [comment]
@rdiachenko https://github.com/checkstyle/checkstyle/issues/7739#issuecomment-611372244 this comment provide reasons.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java [hunk 1]
```diff
@@ -90,6 +87,9 @@
  * JavadocBlockTagLocation
  * </li>
  * <li>
+ * JavadocMethod
+ * </li>
+ * <li>
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java:92]: Please clarify why did you move it here?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 2]
```diff
@@ -721,16 +721,20 @@ private static int calculateTagColumn(Matcher javadocTagMatcher,
...
+            DetailAST exception = throwsAST.getFirstChild();
+            while (exception != null) {
+                if (exception.getType() == TokenTypes.IDENT
+                        || exception.getType() == TokenTypes.DOT) {
+                    DetailAST child = exception;
                     final FullIdent ident = FullIdent.createFullIdent(child);
-                    final ExceptionInfo exceptionInfo = new ExceptionInfo(
+                    while (child.getType() == TokenTypes.DOT) {
+                        child = child.getFirstChild();
+                    }
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:732]: Please explain why do you need this logic?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 3]
```diff
@@ -721,16 +721,20 @@ private static int calculateTagColumn(Matcher javadocTagMatcher,
         final DetailAST throwsAST = ast
                 .findFirstToken(TokenTypes.LITERAL_THROWS);
         if (throwsAST != null) {
-            DetailAST child = throwsAST.getFirstChild();
-            while (child != null) {
-                if (child.getType() == TokenTypes.IDENT
-                        || child.getType() == TokenTypes.DOT) {
+            DetailAST exception = throwsAST.getFirstChild();
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:724]: Please don't rename variables in scope of one update, it is hard to see what was actually changed/added.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 4]
```diff
@@ -752,8 +756,12 @@ private static int calculateTagColumn(Matcher javadocTagMatcher,
                 if (!isInIgnoreBlock(blockAst, throwAst)) {
                     final DetailAST newAst = throwAst.getFirstChild().getFirstChild();
                     if (newAst.getType() == TokenTypes.LITERAL_NEW) {
-                        final FullIdent ident = FullIdent.createFullIdent(newAst.getFirstChild());
-                        final ExceptionInfo exceptionInfo = new ExceptionInfo(
+                        DetailAST child = newAst.getFirstChild();
+                        final FullIdent ident = FullIdent.createFullIdent(child);
+                        while (child.getType() == TokenTypes.DOT) {
+                            child = child.getFirstChild();
+                        }
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:763]: Same question as above, what's the purpose of this `while`?

**Comment by romani** [comment]
@HuGanghui , please rebase to resolve conflicts.
@rdiachenko , please use "Request changes" in review to make it clear your are blocking PR and and answers are required.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java (hunk 1 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java:83]: reason is the 1. I refer to PR #8214 which add it here. 2. As mentioned in https://github.com/checkstyle/checkstyle/issues/7739#issuecomment-611372244 the 1,3,6th log within the check is logging the tags inside the javadoc block comment. It is not accessible using DetailAST. Maybe this is reason to add to here.
https://github.com/checkstyle/checkstyle/blob/d48b815db7c6e9a9272ec78fc7d2cec0d53a8e26/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java#L74-L114

Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java:80]: I don't understand your response. Here, you put `<li>JavadocMethod</li>` after `<li>JavadocBlockTagLocation</li>`. But before your changes the `<li>JavadocMethod</li>` was few lines upper, after `<li>Indentation</li>`. So, what's the reason for this movement?

Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java:80]: Ok, I got it. The `<ul>` element is now different.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 2 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:732]: this is because there is a special case:
https://github.com/HuGanghui/checkstyle/blob/f88da2028c8167f4b7b7ef7f9fd6350a5f69b5ec/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodExtraThrows.java#L77-L82

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 4 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:763]: same special case mentioned above~

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 3 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: done, recover to original name

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 4 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:763]: It looks like desire to keep the same column for violation.

```
$ cat Test.java 
class Test {
    public void bar() {
        throw new org.apache.tools.ant.BuildException(""); //warn
    }
}

$java -jar /var/tmp/checkstyle-8.32-all.jar -t Test.java 
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> Test [1:6]
`--OBJBLOCK -> OBJBLOCK [1:11]
    |--LCURLY -> { [1:11]
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |   |--MODIFIERS -> MODIFIERS [2:4]
    |   |   `--LITERAL_PUBLIC -> public [2:4]
    |   |--TYPE -> TYPE [2:11]
    |   |   `--LITERAL_VOID -> void [2:11]
    |   |--IDENT -> bar [2:16]
    |   |--LPAREN -> ( [2:19]
    |   |--PARAMETERS -> PARAMETERS [2:20]
    |   |--RPAREN -> ) [2:20]
    |   `--SLIST -> { [2:22]
    |       |--LITERAL_THROW -> throw [3:8]
    |       |   |--EXPR -> EXPR [3:14]
    |       |   |   `--LITERAL_NEW -> new [3:14]
    |       |   |       |--DOT -> . [3:38]
    |       |   |       |   |--DOT -> . [3:34]
    |       |   |       |   |   |--DOT -> . [3:28]
    |       |   |       |   |   |   |--DOT -> . [3:21]
    |       |   |       |   |   |   |   |--IDENT -> org [3:18]
    |       |   |       |   |   |   |   `--IDENT -> apache [3:22]
    |       |   |       |   |   |   `--IDENT -> tools [3:29]
    |       |   |       |   |   `--IDENT -> ant [3:35]
    |       |   |       |   `--IDENT -> BuildException [3:39]
    |       |   |       |--LPAREN -> ( [3:53]
    |       |   |       |--ELIST -> ELIST [3:54]
    |       |   |       |   `--EXPR -> EXPR [3:54]
    |       |   |       |       `--STRING_LITERAL -> "" [3:54]
    |       |   |       `--RPAREN -> ) [3:56]
    |       |   `--SEMI -> ; [3:57]
    |       `--RCURLY -> } [4:4]
    `--RCURLY -> } [5:0]
```
@rdiachenko ,  it is hard choice on where to put violation `org [3:18]` vs `BuildException [3:39]`

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:763]: I agree to keep column unchanged.
@HuGanghui , please make a static private method to get node where class name starts. Please generate diff regression report, as code change is significant.

Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: done, diff report will update later

**Comment by rdiachenko** [comment]
@HuGanghui could you provide a config you used for diff report generation?

**Comment by HuGanghui** [comment]
@rdiachenko config is here:
```xml
<module name="JavadocMethod">
            <property name="id" value="default"/>
        </module>

        <module name="JavadocMethod">
            <property name="id" value="change1"/>
            <property name="scope" value="public"/>
            <property name="allowMissingParamTags" value="true"/>
        </module>

        <module name="JavadocMethod">
            <property name="id" value="change2"/>
            <property name="scope" value="private"/>
            <property name="excludeScope" value="protected"/>
        </module>

        <module name="JavadocMethod">
            <property name="id" value="change3"/>
            <property name="validateThrows" value="true"/>
        </module>

        <module name="JavadocMethod">
            <property name="id" value="change4"/>
            <property name="scope" value="public"/>
            <property name="allowMissingParamTags" value="true"/>
        </module>
```

**Comment by timurt** [comment]
@rdiachenko looks good
@HuGanghui please rebase on latest master and we ready to merge

**Comment by HuGanghui** [comment]
@timurt have rebased

**Comment by HuGanghui** [comment]
travis failure reason is :

image

**Comment by rdiachenko** [comment]
I restarted the travis build.
